### ISSUE TITLE
@tables vs @indexes

### ISSUE THREAD

**Issue by konsumer** [architect/arc.codes#155]
What is the difference between the 2? When should I use `@indexes`, as `@tables`seems to do the same thing (define indexes) and seems required for `@indexes` to work?

**Comment by brianleroux** [comment]
tables define dymamodb tables and indexes define indexes on a dynamodb table. a dynamodb table has a partition key (and an optional sort key) and so does an index (which are usually different because you want to query the table from different data). LMK if this answers ur question!

**Comment by konsumer** [comment]
That's confirms my understanding of the arc file, which is why I ask. Here is an example, how is this different?

```
@tables
people
  pplID *String
  insert Lambda
  update Lambda
  delete Lambda

cats
  pplID *String
  catID **String
```
vs

```
@tables
people
cats

@indexes
people
  pplID *String
  insert Lambda
  update Lambda
  delete Lambda

cats
  pplID *String
  catID **String
```

Since I need to add the def for table/indexes twice, it seems redundant (I would just use the first one.)

Ah! does this describe [GSI](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/GSI.html)? That isn't at all clear to me, from the docs. should I make a PR?

**Comment by konsumer** [comment]
What if there was another syntax for GSI, to make it more compact and easier to read?

```
@tables
people
  pplID *String
  insert Lambda
  update Lambda
  delete Lambda
  faceShape **StringGSI

cats
  pplID *String
  catID **String
```

**Comment by brianleroux** [comment]
we can def think on additional/different grammars … as far as the architect/parser is concerned everything comes back as as primitive JSON values. idk that we want to introduce breaking changes but very open to additive "change". 

one thing to note is in Architect 6.x we removed the `Lambda` directive and instead just have one way to do DynamoDB streams. concretely based on your first example this is correct syntax/grammar today:

```
@tables
people
  pplID *String
  stream true

cats
  pplID *String
  catID **String
```

**Comment by konsumer** [comment]
Makes sense. Thanks for the explanation.

**Comment by konsumer** [comment]
I am currently evaluating arc, and seeing if it'll work for what I'm doing (seems awesome, btw.) If I end up using it, I'd be happy to make an additive-only change, if there is interest, to the syntax (like maybe just parse & move stuff with `+` to `@index`, and replace `+` with `*`.)

> we removed the Lambda

Ah, good to know. I pulled that example from the docs on main page, so maybe that should also be updated.

**Comment by konsumer** [comment]
I amy be misunderstanding, and forgive me if this is the wrong place to ask, but it seems like the old syntax would allow me to hook lambdas into insert/update/delete events from dynamodb. How would I do this with the new syntax?

**Comment by brianleroux** [comment]
the new syntax is more true to the implementation; you only get one dep free lambda for all stream events and then its up to your logic to filter on the events you're interested in. we have helper functions in `@architect/functions` library (but folks wanted a dep free default instead of our sugar)

**Comment by konsumer** [comment]
I am still struggling with how to use these. I don't see any reference for using the indexes, once they are created, and I tried logging the `table` object, but it just has these, nothing about indexes:

```
  delete: [Function: _promisified],
  get: [Function: _promisified],
  put: [Function: _promisified],
  query: [Function: _promisified],
  scan: [Function: _promisified],
  update: [Function: _promisified]
```

I have an .arc like this:

```
@tables
experiments
  id: *String
  owner **String
variants
  id: *String
  owner **String

@indexes
experiments
  running *Number
  start **Number
variants
  experiment *String
```

which I intend to describe this:

* I don't want to scan, I want to pre-index by values and query that, for efficiency
* `experiments` & `variants` have unique primary key `id`
* `experiments` & `variants` have non-unique sort key `owner`
* I will want to query `experiments` by `running` & `start` (timestamp) as well as `owner` and `id`
* I will want to query `variant` by `experiment` (`id` of experiment) as well as `owner` and `id`

My queries in English would be this:

* "Give me all experiments with this owner, and running=1"
* "Give me all experiments with start between these 2 numbers, with this owner, and running=0"
* "Give me all the variants for this experiment and this owner"
* "Give me this one experiment, by id, with this owner"
* "Give me this one variant, by id, with this owner"

First, does this describe this correctly, and is my way of modeling the problem correct? And second, how do I query this? I tried this for 1st English query, above:

```
    const results = await tables.experiments.query({
      KeyConditionExpression: '#owner = :owner AND #running = :running',
      ExpressionAttributeNames: {
        '#owner': 'owner',
        '#running': 'running'
      },
      ExpressionAttributeValues: {
        ':owner': user.sub,
        ':running': 1
      }
    })
```
And I get `Query condition missed key schema element: id`.

**Comment by konsumer** [comment]
I noticed `await require('@architect/functions').tables()` has an async `reflect()` method, which seems to do what the docs talk about `_name` (which doesn't seem to exist anymore) and I can get the full names of  the tables. I also need the names of the index, though, to actually use them, right?

**Comment by defionscode** [comment]
@konsumer GSIs aren't exposed as 'new' things. They allow you query the same 'Table' via a totally different partiion (and, optional sort key).

I highly recommend watching this https://www.youtube.com/watch?v=HaEPXoXVf2k and really digging through the aws dynamodb docs. It wasn't until I took the time to do that that I was able to feel comfortable with the DynamoDB-isms.

One of the things that AWS-proper consistently recommends is having a single table for a given application (with some exceptions). 

What you want to do, seems it can fit into the single table pattern.

 I think you'll find you might be overcomplicating your design due to looking at it from SQL syntax perspective

**Comment by konsumer** [comment]
> I highly recommend watching this

I'm comfortable with nosql in general, dynamodb, and node aws-sdk lib, more specifically. I have seen that video before, but rewatched it, just now, to make sure I'm not missing anything big.

My problem here is that arc doesn't seem to follow the aws-sdk index-stuff, and I'm trying to translate. I'm still not clear how to use the sort of indexes I want, and I can't find a straightforward way to create & use LSI and/or multiple GSI.

I guess my question kinda devolved into a few areas, but could be summarized like this: "I know how to use lambdas and dynamo the regular way, but am struggling to map this to arc, how should I do this?"

> I think you'll find you might be overcomplicating your design due to looking at it from SQL syntax perspective

I used SQL syntax to try to make it clearer what I am roughly trying to accomplish with arc. I include the queries because it's a succinct (albeit simplified) way to show some of my application's access patterns (the corresponding JSON dynamo queries & model-definitions are more verbose and hard to read.) I like the tables the guy used in the video mapping RPCs + inputs to indexes, though, so I might map all my graphql queries & relationship-points into something like that.  I feel like GraphQL SDL is already documenting those mappings & relationships mostly for me, but it might make things clearer on the dynamo-side, and it's even more succinct for explaining what I want to do with arc.

In this situation it would greatly simplify things if the GraphQL server on top of this didn't need to do stuff with users, variants & experiments, separate from their parent-model, with or without a reference to the parent-id or grandparent-id. I think this isn't "SQL guy wants normalization, which isn't how NoSQL works." I think maybe the thing that is complicating things here isn't my mental-model, but instead my application's access patterns combined with me not understanding how indexes & lambdas work in arc (which seems different from aws-sdk, and even different from what is in the published arc docs, in some areas.)

> What you want to do, seems it can fit into the single table pattern.

The idea that I could use a single top-level "user" or "experiment" object & some pre-computed indexes (or views in couchdb) to get the data I need, was mind-blowing when I was first learning about NoSQL, years ago. Alternatively, ramming multiple types into the same table and using partition keys to keep it all together on-disk, across a huge cluster, is a good approach for spreading and partitioning data efficiently (he talks a lot about that in the video.) My issue isn't really either of these, though, it's more specific to arc.

I think these things are covered in the docs:

* partition & sort keys: is this done with `*` and `**` in `@tables`
* GSI is this done with `*` (partition) and `**` (sort) in `@indexes`
* stream lambdas don't work as they are in the arc docs, but seem pretty simple

but I also have these questions:

* Are LSI possible? Only GSI are mentioned in the docs, from what I can tell.
* How do I get the `IndexName` in `query`? If this isn't available, how do you even do a GSI query?
* is there a limit in arc of 1 GSI per table? it sort of seems like it from the `@indexes` syntax, as they are keyed by table-name. If so, how do you facet the indexes in multiple ways (like multiple GSIs for reverse-lookup, in the video)

It might be helpful to me if someone who understands both tried to map existing dynamo/lambda examples directly into arc, as it exists right now. Like I think I could totally follow if all the use-cases in that video (or even better: the official AWS docs) were mapped to arc definitions & code, and it would go a long way towards me understanding how to translate my working code & data into arc.

[Here](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/bp-modeling-nosql-B.html) is a fairly complete example of real-world use-case, that he mentions in the video.

* How would you make this in arc with `@tables` and `@indexes`?
* How would you query all those access-patterns, in actual code?

**Comment by brianleroux** [comment]
Hey @konsumer I am so sorry about the ambiguity in our docs. It is our top priority to fix this year and we're just ramping up on it. Believe it or not the intent here is to make working with DynamoDB easier! And its pretty clear we're not succeeding in that goal. So again: sorry about the state of things here. Now let me try to explain.

We do nothing special with the tables in DynamoDB. You can interact with them directly using the `aws-sdk`. There are three possible states for your table resources by default: working locally, `staging` and `production`. When you're working locally we look at the `.arc` file and generate tables and indexes in a local Dynalite instance on port 5000. You can connect to this local in-memory database and read/write data (its super fast so its great for local dev and test logic). 

When you run `arc deploy` we create a default CloudFormation stack which will generate table names. This is intentional and desirable for Infra as Code. Without going into that let me explain what happens. Your `@app` name converted to CamelCase appended with the string `Staging`.  If your app name is `foo` the deployed stack will be named `FooStaging`. This namespace will prefix all resources generated by the CloudFormation stack including your tables. Running `arc deploy production` will create an identical copy of the stack except intended as the prod env. 

So we obviously can't have generated table names in our runtime code. AWS provides a few solutions for this but the one we're using for Architect is called SSM. We store the human readable name and map it to the CloudFormation generated name for lookup at runtime. `@architect/functions` handles these lookups be it locally, `staging` or `production` transparently. 

> Are LSI possible? Only GSI are mentioned in the docs, from what I can tell.

Yes, but not directly supported (or really recommended). Either way you could add via a macro.

> How do I get the IndexName in query? If this isn't available, how do you even do a GSI query?

The index is a concatenation of the partition key name, the sort key name and `-index`. So given the the following `.arc`

```
@indexes
cats
  personID *String
  age **Number
```
The index would be `personID-age-index`. Seemed like a reasonable convention and we absolutely need to document this better. 

[I created a quick project to illustrate the stuff above here.](https://github.com/brianleroux/arc-example-tables)

Here's the project indexes in the console:
<img width="413" alt="Screen Shot 2020-01-21 at 8 37 32 PM" src="https://user-images.githubusercontent.com/990/72866121-1cd60780-3c8f-11ea-809d-3e09c3e79424.png">


>    is there a limit in arc of 1 GSI per table...

You can have many indexes on one table! =) There is a limit of 20 indexes (the AWS limit). The `.arc` format is [more list based than map based](https://runkit.com/brianleroux/5e27d26b14bf70001a486af7). 

> It might be helpful to me if someone who understands both tried to map existing dynamo/lambda examples directly into arc, as it exists right now. Like I think I could totally follow if all the use-cases in that video (or even better: the official AWS docs) were mapped to arc definitions & code, and it would go a long way towards me understanding how to translate my working code & data into arc.

This is an amazing idea for our docs. Cc @pchinjr @ryanblock

**Comment by konsumer** [comment]
> Hey @konsumer I am so sorry about the ambiguity in our docs.

No prob! First, I totally get just being busy. We're all making software for free, have day jobs & family, etc, and we all have lots of things to work on. Docs often come second to features and fixes, so I completely understand. Not your fault or responsibility. Also, I'd like to help, but my understanding of arc needs to grow a bit I think, and I haven't had time myself to really dig in and figure it out. I have code doing what I need to do with dynamo and lambda already, so when I couldn't work it out with arc in a few hours, I just used that. I fully believe it will make my project's management of cloud resources simpler, and easier to maintain. Ya'll are doing great work, and I hope I didn't come off too negative.

> Believe it or not the intent here is to make working with DynamoDB easier!

Oh no, I see it. arc is much simpler than cloudformation or terraform, by a mile. I am also really drawn to the integrated data-layer, which in many ways simplifies more than zeit now or apex up, especially if I don't need their other features (like using lambdas to build lambdas, etc.) I have recently been pretty into pulumi for IaC and we use it where I work for infra, but arc is much simpler & easier to read than that, too. For most things, I think it will save me a lot of time, once I understand how to use it better. I think even as just a config-format, it might be useful for other things, like I could imagine simply parsing an .arc file using your lib, and generating pulumi config to better integrate with our existing infra, with a nice simple arc-only local dev setup.

> We do nothing special with the tables in DynamoDB. You can interact with them directly using the aws-sdk. 

Yeah, that is sort of what I worked out, but I couldn't figure out how to get the name of the GSI, without going into the AWS web admin. I saw your note on this. Makes perfect sense.

> The index would be personID-age-index. Seemed like a reasonable convention and we absolutely need to document this better.

That makes sense. I can live that for sure. Yes it should be in the docs, but it's a small thing.

> So we obviously can't have generated table names in our runtime code. AWS provides a few solutions for this but the one we're using for Architect is called SSM. We store the human readable name and map it to the CloudFormation generated name for lookup at runtime. @architect/functions handles these lookups be it locally, staging or production transparently.

Ah! I think this was a big missing piece for me. I will dig into this a bit more. I am used to the meta-info just being available (like I inputted it somewhere as config string, or in pulumi it's all managed at the resource-level, so easy to grab.) I like your approach though, I think I just needed more reference. This is another thing I'd be happy to help with.

>> Are LSI possible?
> Yes, but not directly supported (or really recommended). Either way you could add via a macro.

That is fair. I generally use them (carefully) for consistency of derived-data or multiple sort keys for the same primary, but I think I can mostly work around this, and they definitely have a huge list of caveats, and anti-patterns that can creep in. I forgot about arc macros. I need to look into those, as I think it would perfectly handle any edge-cases or missing stuff I can't work out with what's built-in.

> You can have many indexes on one table! =) There is a limit of 20 indexes (the AWS limit). The .arc format is more list based than map based.

Duh! that makes perfect sense. I don't know why this wasn't immediately apparent to me.

>> map existing dynamo/lambda examples directly into arc
> This is an amazing idea for our docs.

As I figure things out, I'd be happy to help, too. I learn best (deeper understanding, and better retention) by looking at working examples while working through a project, and docs to look up params as needed, so I really appreciate you taking the time to make a demo-repo and runkit to explain the things we're talking about. I think I am actually starting to grock arc!

Another idea:

Have different paths for the same outcome. Since there are a few ways to interact with the stuff that arc can work with, offer "I use the web to admin AWS" or "I use the CLI to admin AWS", "I am comfortable with cloudformation" or "I want to just use arc for everything". These could even be tabs on code-examples, like you see with aws-sdk for different programming languages. Like start off with the theory behind migrating existing data to dynamo (like AWS docs do) and map that to creating an .arc file, and deployment & verification, and the process of actually building a data-server on top of it. Then offer links to tutorials on adding a frontend (or frontend-backend, like graphql.) This might be overkill, but I could see it really helping bridge into dev's existing knowledge and workflow. I can start working on an article based on the AWS example, and get some code together while I'm learning. I think this was the missing piece for me, in the docs, that would have made it all make more sense, faster.

**Comment by brianleroux** [comment]
> Ya'll are doing great work, and I hope I didn't come off too negative.

nope! no worries. and thx! these comments really made my day. 🙏

>  I can start working on an article based on the AWS example, and get some code together while I'm learning. I think this was the missing piece for me, in the docs, that would have made it all make more sense, faster.

this would be amazing! super appreciate all the help we can get!

**Comment by konsumer** [comment]
I started work on porting that AWS example, [here](https://github.com/konsumer/arc-example). I still need to write the API wrapper and the actual GraphQL service, but I think it's all scaffolded out, and I described how to put it all together, for the most part.

**Comment by konsumer** [comment]
As a related sidenote, I am running into discrepancies in the docs that I think contributed to my earlier confusion, and point out that things aren't exactly a 1-to-1 relationship between APIs, and slow down the learning process.

Here is an example:

According to the arc docs I should be able to do something like this:

```js
   const t = await arc.tables()
    return t.hroe.query({
      key_condition_expression: 'pk = :employeeId',
      expression_attribute_values: {
        ':employeeId': employeeId
      }
    })
```

This throws query errors about missing `KeyConditionExpression`. So I'm like "oh it must be just using the official [aws-sdk syntax](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_Examples)":

```js
    const t = await arc.tables()
    return t.hroe.query({
      key_condition_expression: 'pk = :employeeId',
      expression_attribute_values: {
        ':employeeId': {"S": employeeId }
      }
    })
```

But this doesn't work either. It throws an error about M-type param, which I gathered is "don't use typed-object param syntax". The working syntax is somewhere between all of these, and makes all the docs not correct:

```js
    const t = await arc.tables()
    return t.hroe.query({
      KeyConditionExpression: 'pk = :employeeId',
      ExpressionAttributeValues: {
        ':employeeId': employeeId
      }
    })
```

Should I make separate issues for all of these that I find? Is someone else working on stuff like this in the docs, right now?

**Comment by konsumer** [comment]
Here is another one:

> The index would be personID-age-index. Seemed like a reasonable convention and we absolutely need to document this better.

This makes total sense, but I get an error with this:

```
@tables
hroe
  pk *String
  sk **String

@indexes
hroe
  gsi1-pk *String
  gsi1-sk **String
```

```js
  // Query Employee Details by Employee Name
  // GSI1-PK="EmployeeName"
  async employeeDetailsByName (employeeName) {
    const t = await (arc.tables()).hroe
    return t.query({
      IndexName: 'gsi1-pk-gsi1-sk-index',
      KeyConditionExpression: 'pk = :employeeName',
      ExpressionAttributeValues: {
        ':employeeName': employeeName
      }
    })
  }
```
I get this::

```
The table does not have the specified index: gsi1-pk-gsi1-sk-index
```

I am guessing dashes are messing with it somehow.

Is there a quick way to lookup what it resolves to, without having to deploy it? Maybe something in `repl` or a util-lib that does this? After I know what it should be named, I can just make a constant for `GSI1`, so the code is a literal translation of the implementation comment.

**Comment by brianleroux** [comment]
The interface in the docs should be right-- it is based on the `aws.DynamoDB.DocumentClient` (and not the `aws.DynamoDB` client). Docs for it are here https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html

in the second comment I think you have a mistake…does this work:

```javascript
 async employeeDetailsByName (employeeName) {
    const data = await arc.tables()
    const t = data.hroe
    return t.query({
      IndexName: 'gsi1-pk-gsi1-sk-index',
      KeyConditionExpression: 'pk = :employeeName',
      ExpressionAttributeValues: {
        ':employeeName': employeeName
      }
    })
  }
```

**Comment by brianleroux** [comment]
(note: I had to edit the await statement and the keycondexp!)

**Comment by konsumer** [comment]
> does this work:

I get this error:

```
Invalid KeyConditionExpression: Syntax error; Expected \"(\", \".\", \"<\", \"<=\", \"<>\", \"=\", \">\", \">=\", \"BETWEEN\", \"IN\", \"[\", \"_\", [0-9], or [a-zA-Z] but \"-\" found.
```

> The interface in the docs should be right-- it is based on the aws.DynamoDB.DocumentClient (and not the aws.DynamoDB client)

Ah, hmm, I guess it's the [AWS docs](https://docs.aws.amazon.com/amazondynamodb/latest/APIReference/API_Query.html#API_Query_Examples) that are confusing. Annoying that the dynamo APIs are so similar, but different. The discrepancy still stands in arc docs, though, underscores vs CamelCase don't work (as they are in the example query.)

**Comment by konsumer** [comment]
When I remove the dashes liek this:

```
@indexes
hroe
  gsi1pk *String
  gsi1sk **String
```

I still get an error with `gsi1pk-gsi1sk-index`:

```
ValidationException: The table does not have the specified index: gsi1pk-gsi1sk-index
```

**Comment by konsumer** [comment]
What's even more baffling is I realized I could lookup the generated stuff by just running `arc package`. Duh! Looking at it, though, It actually creates the GSIs the way I would expect:

```json
{
  "HroeTable": {
      "Type": "AWS::DynamoDB::Table",
      "Properties": {
        "KeySchema": [
          {
            "AttributeName": "p",
            "KeyType": "HASH"
          },
          {
            "AttributeName": "s",
            "KeyType": "RANGE"
          }
        ],
        "AttributeDefinitions": [
          {
            "AttributeName": "p",
            "AttributeType": "S"
          },
          {
            "AttributeName": "s",
            "AttributeType": "S"
          },
          {
            "AttributeName": "ap",
            "AttributeType": "S"
          },
          {
            "AttributeName": "as",
            "AttributeType": "S"
          },
          {
            "AttributeName": "bp",
            "AttributeType": "S"
          },
          {
            "AttributeName": "bs",
            "AttributeType": "S"
          }
        ],
        "BillingMode": "PAY_PER_REQUEST",
        "GlobalSecondaryIndexes": [
          {
            "IndexName": "ap-as-index",
            "KeySchema": [
              {
                "AttributeName": "ap",
                "KeyType": "HASH"
              },
              {
                "AttributeName": "as",
                "KeyType": "RANGE"
              }
            ],
            "Projection": {
              "ProjectionType": "ALL"
            }
          },
          {
            "IndexName": "bp-bs-index",
            "KeySchema": [
              {
                "AttributeName": "bp",
                "KeyType": "HASH"
              },
              {
                "AttributeName": "bs",
                "KeyType": "RANGE"
              }
            ],
            "Projection": {
              "ProjectionType": "ALL"
            }
          }
        ]
      }
    }
}````

**Comment by brianleroux** [comment]
can you login into the console and see whats been provisioned ?

**Comment by konsumer** [comment]
> can you login into the console and see whats been provisioned ?

I'm not sure how to do this. I'm trying to avoid making a bunch of random junk on my work's real cloud infra, so if I can do it locally in the emulator it'd be much better (otherwise I would just deploy and check in web console.)

**Comment by konsumer** [comment]
Here is the sam.json:

[sam.json.txt](https://github.com/architect/arc.codes/files/4106257/sam.json.txt)

Could it be a stale emiulated table?

**Comment by brianleroux** [comment]
it could be. the tables are maybe being generated wrong too. can you run describeTable against one to see what the indexes are ?

**Comment by konsumer** [comment]
> it could be. the tables are maybe being generated wrong too. can you run describeTable against one to see what the indexes are ?

In what context? I can't find `describeTable` in `data` or `data.hroe`, in repl.

**Comment by konsumer** [comment]
Maybe [gitter](https://gitter.im/) or discord would be faster to troubleshoot? I like gitter as it's tied to github projects.

**Comment by konsumer** [comment]
I was playing around with [nosql workbench](https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/workbench.html) and connected to port 5000 to see if I could mess with the arc app.

I started with this arc def:

```
@app
arc-hroe-example

@http
get /
post /

@tables
hroe
  PK *String
  SK **String

@indexes
hroe
  SK *String
  Data **String
hroe
  GSI-Bucket *Number
```

and got this key/name structure, which doesn't seem to match docs or what we are talking about:

```js
// aws dynamodb describe-table --table-name arc-hroe-example-staging-hroe --endpoint-url http://localhost:5000

{
    "Table": {
        "AttributeDefinitions": [
            {
                "AttributeName": "SK",
                "AttributeType": "S"
            },
            {
                "AttributeName": "Data",
                "AttributeType": "S"
            },
            {
                "AttributeName": "GSI-Bucket",
                "AttributeType": "N"
            },
            {
                "AttributeName": "PK",
                "AttributeType": "S"
            },
            {
                "AttributeName": "SK",
                "AttributeType": "S"
            }
        ],
        "TableName": "arc-hroe-example-staging-hroe",
        "KeySchema": [
            {
                "AttributeName": "PK",
                "KeyType": "HASH"
            },
            {
                "AttributeName": "SK",
                "KeyType": "RANGE"
            }
        ],
        "TableStatus": "ACTIVE",
        "CreationDateTime": 1580295250,
        "ProvisionedThroughput": {
            "NumberOfDecreasesToday": 0,
            "ReadCapacityUnits": 5,
            "WriteCapacityUnits": 5
        },
        "TableSizeBytes": 0,
        "ItemCount": 0,
        "TableArn": "arn:aws:dynamodb:us-west-2:000000000000:table/arc-hroe-example-staging-hroe",
        "TableId": "1ae3f229-628a-4a6d-5c6c-d3d3d82a",
        "GlobalSecondaryIndexes": [
            {
                "IndexName": "arc-hroe-example-staging-hroe-SK-Data-index",
                "KeySchema": [
                    {
                        "AttributeName": "SK",
                        "KeyType": "HASH"
                    },
                    {
                        "AttributeName": "Data",
                        "KeyType": "RANGE"
                    }
                ],
                "Projection": {
                    "ProjectionType": "ALL"
                },
                "IndexStatus": "ACTIVE",
                "ProvisionedThroughput": {
                    "NumberOfDecreasesToday": 0,
                    "ReadCapacityUnits": 5,
                    "WriteCapacityUnits": 5
                },
                "IndexSizeBytes": 0,
                "ItemCount": 0,
                "IndexArn": "arn:aws:dynamodb:us-west-2:000000000000:table/arc-hroe-example-staging-hroe/index/arc-hroe-example-staging-hroe-SK-Data-index"
            },
            {
                "IndexName": "arc-hroe-example-staging-hroe-GSI-Bucket-index",
                "KeySchema": [
                    {
                        "AttributeName": "GSI-Bucket",
                        "KeyType": "HASH"
                    }
                ],
                "Projection": {
                    "ProjectionType": "ALL"
                },
                "IndexStatus": "ACTIVE",
                "ProvisionedThroughput": {
                    "NumberOfDecreasesToday": 0,
                    "ReadCapacityUnits": 5,
                    "WriteCapacityUnits": 5
                },
                "IndexSizeBytes": 0,
                "ItemCount": 0,
                "IndexArn": "arn:aws:dynamodb:us-west-2:000000000000:table/arc-hroe-example-staging-hroe/index/arc-hroe-example-staging-hroe-GSI-Bucket-index"
            }
        ]
    }
}
```

**Comment by konsumer** [comment]
Also, these don't match what's in `arc package`:
[sam.json.txt](https://github.com/architect/arc.codes/files/4127599/sam.json.txt)

**Comment by ryanblock** [comment]
@konsumer we're all on [Slack](https://join.slack.com/t/architecture-as-text/shared_invite/MjE2MzU4Nzg0NTY1LTE1MDA2NzgyMzYtODE2NzRkOGRmYw), please do join us there and we'd be happy to dig in with you. May also want to jump onto a Zoom call, since that may be faster still.

Re. the `describe-table` output and SAM JSON you posted today, giving it a once over I'm not seeing any issues but perhaps I missed something. What specifically doesn't look right?

**Comment by konsumer** [comment]
> Re. the describe-table output and SAM JSON you posted today, giving it a once over I'm not seeing any issues but perhaps I missed something. What specifically doesn't look right?

The index-name doesn't match what you said above (it has the project-name and target in it: `arc-hroe-example-staging-hroe-GSI-Bucket-index`, not `GSI-Bucket-index`) and they don't match each other (`arc-hroe-example-staging-hroe-GSI-Bucket-index` actual vs `GSI-Bucket-index`, which is in sam.json.) I replaced `Data` in current code with `SearchData`, as it's a keyword.

Using these known key-names queries started working:

```js
const GSI1 = 'arc-hroe-example-staging-hroe-SK-SearchData-index'
const GSI2 = 'arc-hroe-example-staging-hroe-SearchData-index'
const GSI3 = 'arc-hroe-example-staging-hroe-GSIBucket-index'
```
I will probly work out how to lookup key-names automatically, though, since this will fail on production, or if I change the names of anything.

my point now is that `arc package` & actual aren't agreeing on key-names, which seems very bad. I tried deleting all the tables/indexes/data and start `arc sandbox`, and it keeps creating them in this format.

> @konsumer we're all on Slack, please do join us there and we'd be happy to dig in with you. May also want to jump onto a Zoom call, since that may be faster still.

Nice! I didn't realize there was a slack. I'd be happy to jump on. I can't seem to get Zoom to work on Arch-Linux/Chrome, but I can do some other thing like Discord, Hangouts, Skype, etc. [jitsi](https://jitsi.org/) seems similar to zoom, but is open-source and works fine for me.
